feat: Auth v3 - Enhanced Authentication System & LinkedIn Provider!!!

meczajkowski opened the pull request:
Finally! 

# feat: Auth v3 - Enhanced Authentication System & LinkedIn Provider!!!
### Overview 
This PR significantly upgrades our authentication system by introducing LinkedIn as a provider for Specialists and Hunters. Users can now register and log in using LinkedIn, GitHub, or both. This update also optimizes JWT handling, enhances access control, and adds support for custom slugs in user profiles.

Additionally, database calls within JWT and session callbacks have been minimized. The JWT callback now only queries the database during sign-in, reducing unnecessary calls during the session lifecycle.

### Key Changes

- _**LinkedIn Integration for Specialists and Hunters:**_ Specialists and Hunters can now sign up and log in via LinkedIn. This provides more flexibility, especially for users without GitHub accounts.

- _**Email Provider for Hunters:**_ Email authentication is currently available exclusively for Hunters. However, extending it to Specialists is straightforward

- _**Custom Slug Creation for Profiles:**_ Since LinkedIn doesn’t provide a username, users can now create custom slugs during profile creation. For GitHub users, the slug field in the profile creation form is pre-filled with their GitHub usernamebut remains editable.

- **_Account Merging:_** If a user registers with the same email through both LinkedIn and GitHub, their accounts will be seamlessly merged.

- **_Efficient JWT Handling:_** Database calls in JWT and session callbacks have been minimized. The JWT callback now fetches user data from the database only during sign-in, eliminating hundreds of redundant calls during a session lifecycle.

- **_Discord Notifications:_** 

Sign-up notifications now include the authentication provider (LinkedIn or GitHub), providing better context about the user’s origin. Notifications are also sent if a user attempts to add an unauthorized role, including their identifier for easier tracking.

### Technical Details

- Added `slug `column to `Profile `model in the Prisma schema to store custom slugs.
- Refactored role-based access logic to separate and streamline access controls for Hunters and Specialists, enhancing maintainability and security.


### TODO

- [x] Custom Slug Validation: Implement a slug validation mechanism to prevent duplicate or inappropriate slugs.
- [x] GitHub Username Updates: Ensure that GitHub usernames are updated during sign-in. Currently, they are only set during sign-up and don't reflect changes made on GitHub.
- [x] Specialist Role Validation for Email Provider: Add checks to prevent Specialists from signing in via the email provider, avoiding incorrect welcome emails intended for Hunters.

Sumick commented on apps/platform-app/src/app/[locale]/(auth)/(components)/LinkedInLoginButton/LinkedInLoginButton.tsx:
```diff
@@ -0,0 +1,33 @@
+
+interface LinkedInLoginButtonProps {
+  label?: string
+  role: 'SPECIALIST' | 'HUNTER'
```
**nit:** We should have enum for that, eg: `enum Roles` to prevent string duplication

meczajkowski replied:
I have created Roles enum inside of ```src\app\[locale]\(auth)\_models\User.model.ts``` because we can not use Role enum from prisma here

Sumick commented on apps/platform-app/src/app/[locale]/(auth)/(routes)/oauth/page.tsx:
```diff
@@ -0,0 +1,69 @@
+  await sendWelcomeEmail(user.email, devName)
+  await sendDiscordNotificationToModeratorChannel(
+    `User **${devName}** has created an account with **${provider.toUpperCase()}** as ${
+      isSpecialist
```
**nit:** Instead of hardcoding condition we can just display added `role` arg in message

Sumick commented on apps/platform-app/src/app/[locale]/(auth)/(routes)/signin/(components)/LoginHunter/LoginHunter.tsx:
```diff
@@ -104,18 +103,22 @@ const LoginHunter = () => {
         </Button>
-      </form>
+        <span>{t('or')}</span>
+        <LinkedInLoginButton label={t('loginWithLinkedin')} role="HUNTER" />
```
**nit:** We should use `Roles.HUNTER`

Sumick commented on apps/platform-app/src/app/[locale]/(profile)/(components)/UserPhotoUploader/UserPhotoUploader.tsx:
```diff
@@ -84,10 +84,13 @@ export const UserPhotoUploader = () => {
-              setImage={setUserImage}
-              showError={setImageUploadError}
-            />
+            {/* TODO LinkedIn photo uploder */}
```
**suggestion:** Or simpler, remove "Import from Github", and let's leave just "Upload my photo".

Sumick commented on apps/platform-app/src/app/[locale]/(profile)/(routes)/my-profile/(components)/CreateProfile/PersonalInfo/PersonalInfo.tsx:
```diff
@@ -50,20 +51,37 @@ const PersonalInfo = () => {
-            dataTestId=""
-            maxLength={30}
+            tooltipText={t('slugTooltip')}
+            onBlur={handleBlur}
```
**issue:** We should validate this field on blur this way, so it checks in DB is such slug is not already taken. This validation is crucial and we should do it sooner than after form submission

Sumick commented on apps/platform-app/src/app/[locale]/(profile)/_actions/mutations/saveMyProfile.ts:
```diff
@@ -20,6 +20,7 @@ const profilePendingFields: EditProfileFormFields[] = [
   'fullName',
   'email',
   'avatarUrl',
+  // 'slug',
```
**nit:** Comment to remove?

meczajkowski replied:
this is added to remember to discuss if changed slug should change profile state to PENDING or not

Sumick commented on apps/platform-app/src/app/[locale]/(profile)/_actions/queries/findProfileBySlug.ts:
```diff
@@ -0,0 +1,17 @@
+import { withSentry } from '@/utils/errHandling'
+import { type Profile } from '@prisma/client'
+
+export const findProfileBySlug = withSentry(async (slug: Profile['slug']) => {
```
**nit:** Why not just `slug: string`?

meczajkowski replied:
Ofc at the end Profile['slug'] is just a string but this way it's clear that it comes from Profile.

Sumick commented on apps/platform-app/prisma/schema.prisma:
```diff
@@ -56,6 +56,7 @@ model VerificationToken {
 
 model Profile {
   id                       String            @id @default(uuid())
+  slug                     String            @unique
```
**issue:** Slug must be optional, we won't be able to apply such migration

meczajkowski replied:
this would be handled by #369

Sumick reviewed (changes requested):
Found 2 issues to resolve + minor nits

meczajkowski commented:
- Removed autofilling of the slug input with the GitHub username from the session, as there was no proper way to display user information about invalid or taken slugs.
- Added a slug validation mechanism to prevent duplicate or inappropriate slugs.
- Updated GitHub Username handling: Ensured that GitHub usernames are updated during sign-in, as they were previously only set during sign-up and did not reflect changes made on GitHub.
- Specialist Role Validation for Email Provider: Implemented checks to prevent Specialists from signing in via the email provider, avoiding incorrect welcome emails that were intended for Hunters.

NerdbordMentor commented on apps/platform-app/prisma/migrations/20241203175027_change_slug_to_required/migration.sql:
```diff
@@ -0,0 +1,8 @@
+/*
+  Warnings:
+
+  - Made the column `slug` on table `Profile` required. This step will fail if there are existing NULL values in that column.
```
**issue**: Ensure there are no NULL values in the `slug` column before applying this migration to prevent failures.

NerdbordMentor commented on apps/platform-app/src/app/[locale]/(auth)/(components)/GithubLoginButton/GithubLoginButton.tsx:
```diff
@@ -4,10 +4,12 @@ import { Button } from '@gdh/ui-system'
 import { useState } from 'react'
+import { type Roles } from '../../_models/User.model'
 import styles from './GithubLoginButton.module.scss'
 
```
**praise**: Including role in the props enhances the component's flexibility and reusability.

NerdbordMentor commented on apps/platform-app/src/app/[locale]/(auth)/(components)/LinkedInLoginButton/LinkedInLoginButton.tsx:
```diff
@@ -0,0 +1,34 @@
+  role: Roles.SPECIALIST | Roles.HUNTER
+}
+
+export const LinkedInLoginButton = (props: LinkedInLoginButtonProps) => {
```
**praise**: Well-structured component for LinkedIn login, maintaining consistency with GitHub login.

NerdbordMentor commented on apps/platform-app/src/app/[locale]/(auth)/(routes)/signin/(components)/LoginDev/LoginDev.tsx:
```diff
@@ -1,16 +1,30 @@
       <h2>{t('title')}</h2>
       <p>{t('description')}</p>
-      <GithubLoginButton label={t('loginWithGithub')} />
+      <div className={styles.btnsContainer}>
```
**praise**: Nice use of a container for login buttons, improving layout and user experience.

NerdbordMentor commented on apps/platform-app/src/app/[locale]/(auth)/_actions/queries/findUserByEmail.ts:
```diff
@@ -0,0 +1,17 @@
+import { withSentry } from '@/utils/errHandling'
+import { cache } from 'react'
+
+export const findUserByEmail = cache(
```
**praise**: Good use of caching with `cache` to optimize repeated queries.

NerdbordMentor commented on apps/platform-app/src/app/[locale]/(auth)/_models/User.model.ts:
```diff
@@ -1,19 +1,33 @@
   }
 }
+
+export enum Roles {
```
**praise**: Defining roles as an enum improves type safety and reduces errors.

NerdbordMentor commented on apps/platform-app/src/app/[locale]/(auth)/auth.helpers.ts:
```diff
@@ -13,17 +14,32 @@ export const authorizeUser = async () => {
+  return {
+    user: session.user,
+    userIsHunter,
+    userIsSpecialist,
```
**suggestion**: Consider using a more descriptive name for `userHasRole` to clarify its purpose.

NerdbordMentor commented on apps/platform-app/src/app/[locale]/(moderation)/_actions/mutations/approveProfile.ts:
```diff
@@ -2,10 +2,7 @@
 import { getAuthorizedUser } from '@/app/[locale]/(auth)/auth.helpers'
 import { createProfileModel } from '@/app/[locale]/(profile)/_models/profile.model'
 import { sendProfileApprovedEmail } from '@/backend/mailing/mailing.service'
```
**suggestion**: Remove unused imports to keep the code clean and maintainable.

NerdbordMentor commented on apps/platform-app/src/app/[locale]/(profile)/(components)/CVuploader/CvUploader.tsx:
```diff
@@ -1,12 +1,11 @@
-import { CVuploaderForm } from "./CvUploaderForm"
 import styles from './CvUploader.module.scss'
+import { CVuploaderForm } from './CvUploaderForm'
```
**nitpick**: Import statements should be alphabetically ordered for consistency.

NerdbordMentor commented on apps/platform-app/src/app/[locale]/(profile)/(components)/UserPhotoUploader/UserPhotoUploader.tsx:
```diff
@@ -7,7 +7,6 @@ import { ErrorIcon } from '@gdh/ui-system/icons'
 import { useTranslations } from 'next-intl'
 import React, { useEffect, useState } from 'react'
-import GithubUserPhotoUploader from './GithubUserPhotoUploader'
 import styles from './UserPhotoUploader.module.scss'
```
**suggestion**: Consider using a default avatar URL if `session?.user?.avatarUrl` is not available.

NerdbordMentor reviewed (changes requested):
The code changes introduce LinkedIn as an authentication provider, which aligns with the task requirements. The addition of the `LinkedInLoginButton` component and the modifications to the `GithubLoginButton` to include role parameters are well-implemented. The migration to make the `slug` column required in the `Profile` table and the corresponding Prisma schema changes are correctly executed. The refactoring of the OAuth handling to include role assignment and the removal of the `github-oauth` page in favor of a more generic `oauth` page is a positive change.

However, there are several critical issues that prevent approval:

### Things Missing:

- [x] The code does not address the requirement for email authentication for Hunters, with the potential to extend it to Specialists.
- [x] There is no implementation for merging accounts if a user registers with the same email through both LinkedIn and GitHub.
- [x] JWT handling optimization to minimize database calls is not evident in the changes.
- [ ] Custom slug validation to prevent duplicates or inappropriate slugs is missing.
- [ ] Specialist role validation to prevent email provider sign-ins for Specialists is not implemented.

### Things That Went Well:

- The LinkedIn authentication integration is well-executed, with appropriate UI updates.
- The migration and schema changes for the `slug` field are correctly implemented.
- The refactoring of role-based access logic and OAuth handling is a positive improvement.

Sumick reviewed (approved):
Let's get this badboy on staging
